cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewManager.java

```diff
@@ -154,32 +157,66 @@ public void onRemoval(RemovalNotification<ApplicationId, Injector> notification)
             if (runner instanceof Service) {
               stopQuietly((Service) runner);
             }
+            removePreviewDir(injector.getInstance(Key.get(ProgramId.class,
+                                                          Names.named(PreviewRunnerModule.PREVIEW_PROGRAM_ID))));
           }
-          ApplicationId application = notification.getKey();
-          if (application == null) {
-            return;
-          }
-          removePreviewDir(application);
         }
       })
       .build();
   }
 
+  @Override
+  protected void startUp() throws Exception {
+    File previewDir = previewDataDir.toFile();
+
+    File[] files = previewDir.listFiles();
```

Comment:
Use `DirUtils.listFiles` instead. It returns a collection and is guaranteed that it won't be null.

Reply:
fixed